xds/src/main/java/io/grpc/xds/FilterChainMatchingHandler.java

```diff
@@ -0,0 +1,338 @@
+/*
+ * Copyright 2021 The gRPC Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.grpc.xds;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+import static io.grpc.xds.XdsServerWrapper.ATTR_SERVER_ROUTING_CONFIG;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.Strings;
+import com.google.common.collect.Iterables;
+import com.google.protobuf.UInt32Value;
+import io.grpc.Attributes;
+import io.grpc.netty.GrpcHttp2ConnectionHandler;
+import io.grpc.netty.InternalProtocolNegotiationEvent;
+import io.grpc.netty.InternalProtocolNegotiator.ProtocolNegotiator;
+import io.grpc.netty.ProtocolNegotiationEvent;
+import io.grpc.xds.EnvoyServerProtoData.CidrRange;
+import io.grpc.xds.EnvoyServerProtoData.ConnectionSourceType;
+import io.grpc.xds.EnvoyServerProtoData.FilterChain;
+import io.grpc.xds.EnvoyServerProtoData.FilterChainMatch;
+import io.grpc.xds.XdsServerWrapper.ServerRoutingConfig;
+import io.grpc.xds.internal.Matchers.CidrMatcher;
+import io.grpc.xds.internal.sds.SdsProtocolNegotiators.ServerSdsHandler;
+import io.grpc.xds.internal.sds.SslContextProviderSupplier;
+import io.netty.channel.ChannelHandler;
+import io.netty.channel.ChannelHandlerContext;
+import io.netty.channel.ChannelInboundHandlerAdapter;
+import java.net.Inet6Address;
+import java.net.InetAddress;
+import java.net.InetSocketAddress;
+import java.security.cert.CertStoreException;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import javax.annotation.Nullable;
+
+
+/**
+ * Handles L4 filter chain match for the connection based on the xds configuration.
+ * */
+public final class FilterChainMatchingHandler extends ChannelInboundHandlerAdapter {
+  private final GrpcHttp2ConnectionHandler grpcHandler;
+  private final FilterChainSelector selector;
+  @Nullable
+  private final ProtocolNegotiator fallbackProtocolNegotiator;
+
+  /**
+   * Selects the filter chain using the selector configuration.
+   * */
+  public FilterChainMatchingHandler(
+          GrpcHttp2ConnectionHandler grpcHandler, FilterChainSelector selector,
+          @Nullable ProtocolNegotiator fallbackProtocolNegotiator) {
```

Comment:
> We should leave any SslContext logic in the existing "SDS" negotiatior.

I do not mind putting the filter chain selecting logic (what this handler does perfectly matching its name) in existing `SDS` negotiator.
But note that the existing `SDS` negotiator doesn't do that here. It callback a xdsclientWrapperForServerSds to do the sslContext selecting, by injecting that client in attributes.  So the current SslContext logic is somewhere in io.grpc.xds.

>  Instead of this being a fallback, this should be a "delegate" and should be used with the replace() unconditionally (excepting when this closes the connection). The existing sds negotiator would be modified to look up config in the attributes.

The existing sds negotiator just fetches the attribute and constructs another handler that does the filter chain selection. We didn't do much refactoring here. Do you mean it should do more than the construction of a handler the does the selection, but does the selection logic inside SDS negotiator?
As for "delegate" and replace(), creating the next handler based on the selection result (SslContextProviderSupplier-based or fallbackNegotiator) seems just two lines, if we keep using [sdsServerHandler](https://github.com/grpc/grpc-java/blob/d8d378454f0d1f76efd29cb848454f1cde5e714e/xds/src/main/java/io/grpc/xds/internal/sds/SdsProtocolNegotiators.java#L321) , then we do not necessarily need a `delegate` just does the two line logic?

Reply:
> I do not mind putting the filter chain selecting logic (what this handler does perfectly matching its name) in existing SDS negotiator.

No, I do want that here. This code should decide the filter chain to be used, and the SDS code should use that chain's configuration.

> The existing sds negotiator just fetches the attribute and constructs another handler that does the filter chain selection. We didn't do much refactoring here...

We need two partially independent pieces of code. XdsServer needs to choose the filter chain, and XdsCredentials needs to use the filter chain config or use its fallback. Both pieces of code need to work without the other, and the user will be configuring which is used so neither can integrate the code of the other. We _must not_ care about SSL here as that _must only_ be done if using XdsCredentials (which use the SDS code). It isn't about lines of code, but "we must only run certain code in certain circumstances." The choice of using the xds-provided config or using the fallback must only be triggered by XdsCredentials.

Reply:
Looking at the code currently, `ATTR_FILTER_CHAIN_SELECTOR_REF` is consumed by `ServerSdsProtocolNegotiator` which is engaged by `SdsProtocolNegotiators.ServerFactory` which is used only when `XdsServerCredentials` are in effect. Whereas for other use-cases (routing etc) we still need the filter selection logic so we need to find a way to delink this from `XdsServerCredentials`. If that's going to be done later some explanation/TODO comments will help.

Reply:
https://github.com/grpc/grpc-java/pull/8333 intends to implement it.